File: ref-impl/ts/test_dtxt.ts
import * as dtxt from './dtxt';
import * as assert from 'assert';

function testSpecExample() {
    const specExample = `
# DTXT example
{
  name: \`Sample\`,
  created: D(2026-01-15),
  updated: D(2026-01-15T10:30:00Z),
  active: T,
  count: 42,
  big: BN(9007199254740993),
  hash: B(A7B2319E44CE12BA),
  items: [1, 2, 3],
  meta: {
    retries: 3,
    enabled: F,
  },
}
`;
    const parsed = dtxt.parse(specExample);
    console.log("Parsed Example Successfully");

    assert.strictEqual(parsed.name, 'Sample');
    assert.strictEqual(parsed.active, true);
    assert.strictEqual(parsed.count, 42);
    assert.strictEqual(parsed.big, 9007199254740993n);
    assert.deepStrictEqual(parsed.hash, new Uint8Array([0xA7, 0xB2, 0x31, 0x9E, 0x44, 0xCE, 0x12, 0xBA]));
    assert.deepStrictEqual(parsed.items, [1, 2, 3]);
    assert.strictEqual((parsed.meta as any).retries, 3);
    assert.strictEqual((parsed.meta as any).enabled, false);

    // Round trip
    const dumped = dtxt.stringify(parsed);
    console.log("Dumped (Canonical):", dumped);

    const reparsed = dtxt.parse(dumped);
    assert.strictEqual(reparsed.name, parsed.name);
    assert.strictEqual(reparsed.big, parsed.big);
    assert.deepStrictEqual(reparsed.hash, parsed.hash);
    console.log("Round trip successful");
}

function testErrorHandling() {
    try {
        dtxt.parse("{ user.name: 1 }");
        assert.fail("Should have failed on dot in key");
    } catch (e: any) {
        console.log("Caught expected error for dot in key:", e.message);
    }
}

testSpecExample();
testErrorHandling();
console.log("All TypeScript tests passed!");

File: ref-impl/python/tests/test_dtxt.py
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

import dtxt
from datetime import datetime, date
import binascii

def test_spec_example():
    spec_example = """
# DTXT example
{
  name: `Sample`,
  created: Date(2026-01-15),
  updated: Date(2026-01-15T10:30:00Z),
  active: T,
  count: 42,
  big: BigNumber(9007199254740993),
  hash: Binary(A7B2319E44CE12BA),
  items: [1, 2, 3],
  meta: {
    retries: 3,
    enabled: F,
  },
}
"""
    parsed = dtxt.loads(spec_example)
    print("Parsed Example Successfully")
    
    assert parsed['name'] == 'Sample'
    assert parsed['created'] == date(2026, 1, 15)
    # Check updated (datetime with UTC)
    # datetime.fromisoformat might return +00:00 for Z if correctly handled
    assert parsed['active'] is True
    assert parsed['count'] == 42
    assert parsed['big'] == 9007199254740993
    assert parsed['hash'] == binascii.unhexlify('A7B2319E44CE12BA')
    assert parsed['items'] == [1, 2, 3]
    assert parsed['meta']['retries'] == 3
    assert parsed['meta']['enabled'] is False

    # Round trip
    dumped = dtxt.dumps(parsed)
    print("Dumped (Canonical):", dumped)
    
    reparsed = dtxt.loads(dumped)
    assert reparsed == parsed
    print("Round trip successful")

def test_comments_and_whitespace():
    dtxt_text = "{ # comment\n  a: 1, \n /* not supported */ \n b: 2 }"
    # The lexer should fail on /* if not supported, but the spec only mentions //
    # My lexer will treat /* as mismatch if it's not handled.
    try:
        dtxt.loads(dtxt_text)
    except dtxt.DTXTError as e:
        print("Expected error for /*:", e)
        assert "Unexpected character" in str(e)

def test_keys():
    dtxt_text = "{ 123key: T, _key: F, Key9: N }"
    parsed = dtxt.loads(dtxt_text)
    assert parsed['123key'] is True
    assert parsed['_key'] is False
    assert parsed['Key9'] is None

if __name__ == "__main__":
    test_spec_example()
    test_comments_and_whitespace()
    test_keys()
    print("All tests passed!")
